ggshield/verticals/secret/secret_scanner.py

```diff
@@ -70,6 +70,16 @@ def __init__(
             if TokenScope.INCIDENTS_READ not in response.scopes:
                 raise MissingScopesError([TokenScope.INCIDENTS_READ])
 
+        if secret_config.source_uuid:
+            response = self.client.api_tokens()
+
+            if not isinstance(response, (Detail, APITokensResponse)):
+                raise UnexpectedError("Unexpected api_tokens response")
+            elif isinstance(response, Detail):
+                raise UnexpectedError(response.detail)
+            if TokenScope.SCAN_CREATE_INCIDENTS not in response.scopes:
+                raise MissingScopesError([TokenScope.SCAN_CREATE_INCIDENTS])
```

Comment:
In certain situations, for example when scanning commit ranges, ggshield creates multiple SecretScanner instances. This is why we have the `if check_api_key:` at the start of the method: we only check the API key validity if it has not already be done by the caller.

I think these new checks should be done in `check_client_api_key()`. It does not have access to the config though, so you need to extend the function to accept a `source_uuid` optional argument.